adr: Un-Ordered Transaction Inclusion (#18553)
cosmos/cosmos-sdk: docs/architecture/adr-069-unordered-account.md

# ADR 069: Un-Ordered Account

## Changelog

* Nov 24, 2023: Initial Draft

## Abstract

Support gaps in account nonce values to support un-ordered(concurrent) transaction inclusion.

## Context

> Right now the transactions from a particular sender must be included in strict order because of the nonce value requirement, which makes it tricky to send many transactions concurrently, for example, when a previous pending transaction fails or timeouts, all the following transactions are all blocked. Relayer and exchanges are some typical examples.
> 
> The main purpose of nonce value is to protect against replay attack, but for that purpose we only need to make sure the nonce is unique, so we can relax the orderness requirement without lose the uniqueness, in that way we can improve the user experience of concurrent transaction sending.

Reviewer: ```suggestion
Right now, transactions from a particular sender must be included in strict order
because of nonce value requirements, which makes it tricky to send many transactions
concurrently. For example, when a previous pending transaction fails or timeouts,
all the following transactions are blocked. Relayer and exchanges are some typical
examples.

The main purpose of nonce value is to protect against replay attack, but for that
purpose we only need to make sure the nonce is unique, so we can relax the order
requirement without loss to uniqueness, in that way we can improve the user experience
of concurrent transaction sending.
```

## Decision

> Change the nonce logic to allow gaps to exist, which can be filled by other transactions later, or never filled at all, the prototype implementation use a bitmap to record these gap values.

Reviewer 2: if we allow gaps to exist and ordering here doesnt matter, should we look into making nonces expire. A nonce is accompanied with a blockheight or hash at which the nonce will expire. 

This leads into the direction of not needing nonces at all. Since we move from caring about order and replay to only replay protection then nonces could be simplified.

Author: yeah, the basic idea is to allow "unordered" account with as low as possible overhead.

Author: > should we look into making nonces expire

based on the prototype implementation in this ADR, we can record the block height or time the latest nonce is created, and when it timeouts make all the "gaps" expire at the same time, in this way, the overhead should be minimal.

> Change the nonce logic to allow gaps to exist, which can be filled by other transactions later, or never filled at all, the prototype implementation use a bitmap to record these gap values.

Reviewer: Can you please add more to detail and generally explain the concept of "gaps". I think it's clear if you have context,  but it'll be better for the reader.

Author: Added a bit more in decisions sections, please re-review

It's debatable how we should configure the user accounts, for example, should we change the default behavior directly or let user to turn this feature on explicitly, or should we allow user to set different gap capacity for different accounts.

```golang

> const MaxGap = 1024
> 
>   type Account struct {
>     ...
>     SequenceNumber int64
> +   Gaps *IntSet
>   }

Reviewer: ```suggestion
const MaxGap = 1024

type Account struct {
    // ...
    SequenceNumber int64
    Gaps *IntSet
  }
```

// CheckNonce checks if the input nonce is valid, if yes, modify internal state.
func(acc *Account) CheckNonce(nonce int64) error {

>   switch {

Reviewer: I would more godocs to each statement to describe what that case is intended to do. It's not immediately clear to me the last two cases, what is going on.

E.g.

- `nonce >= acct.SequenceNumber + 1:` is for a gap to be filled after the current nonce I think? But it's not clear what the loop is intended to do.
- `nonce < acct.SequenceNumber:` is for a gap that is unfilled, or "in the past". But again, not clear what the logic in the case is intended to do.

Author: see if it's better now.

    case nonce == acct.SequenceNumber:
    	return errors.New("nonce is occupied")
    case nonce >= acct.SequenceNumber + 1:
      gaps := nonce - acct.SequenceNumber - 1
      if gaps > MaxGap {
        return errors.New("max gap is exceeded")
      }
    	for i := 0; i < gaps; i++ {
      	acct.Gaps.Add(i + acct.SequenceNumber + 1)
    	}
    	acct.SequenceNumber = nonce
    case nonce < acct.SequenceNumber:
    	if !acct.Gaps.Contains(nonce) {
      	return errors.New("nonce is occupied")
    	}
    	acct.Gaps.Remove(nonce)
  }
  return nil
}
```

Prototype implementation of `IntSet`:

```golang
type IntSet struct {
  capacity int
  bitmap roaringbitmap.BitMap
}

func NewIntSet(capacity int) *IntSet {
  return IntSet{
    capacity: capacity,
    bitmap: *roaringbitmap.New(),
  }
}

func (is *IntSet) Add(n int) {
  if is.bitmap.Length() >= is.capacity {
    // pop the minimal one

>     is.Remove(is.bitmap.Minimal())

Author: when it's removed from the set, it just means the nonce can't be used anymore, if there's a old pending transaction still using that nonce, it won't be accepted anymore, should be fine, like an expiration.

  }
  
  is.bitmap.Add(n)
}

func (is *IntSet) Remove(n int) {
  is.bitmap.Remove(n)
}

func (is *IntSet) Contains(n int) bool {
  return is.bitmap.Contains(n)
}
```

## Status

> Proposed.

Reviewer: nit: Move this section to the top  :)

Author: done

      for i := 0; i < gaps; i++ {
        acct.Gaps.Add(i + u.Sequence + 1)
      }
      // record the latest nonce
    	u.Sequence = nonce
    case nonce < u.Sequence:
      // try to use a gap value
    	if !u.Gaps.Contains(nonce) {
        return errors.New("nonce is occupied")
    	}
    	u.Gaps.Remove(nonce)
  }
  return nil
}

// IntSet is a set of integers with a capacity, when capacity reached, drop the smallest value
type IntSet struct {
  capacity int
  bitmap roaringbitmap.BitMap
}

func NewIntSet(capacity int) *IntSet {
  return IntSet{
    capacity: capacity,
    bitmap: *roaringbitmap.New(),
  }
}

func (is *IntSet) Add(n uint64) {
  if is.bitmap.GetCardinality() >= is.capacity {
    // drop the smallest one
    is.bitmap.Remove(is.bitmap.Minimal())
  }

  is.bitmap.Add(n)
}

// AddRange adds the integers in [rangeStart, rangeEnd) to the bitmap.
func (is *IntSet) AddRange(start, end uint64) {
  n := end - start
  if is.bitmap.GetCardinality() + n > is.capacity {
    // drop the smallest ones
    toDrop := is.bitmap.GetCardinality() + n - is.capacity
    for i:= 0; i<toDrop; i++ {
      is.bitmap.Remove(is.bitmap.Minimal())
    }
  }

  is.bitmap.AddRange(start, end)

> }

Author: it's not the same as the `RemoveRange`.

# ADR 070: Un-Ordered Transaction Inclusion

## Changelog

* Dec 4, 2023: Initial Draft

## Status

Proposed

## Abstract

> We propose a way to do replay-attack protection without enforcing the order of transactions, it don't use nonce at all. In this way we can support un-ordered transaction inclusion.

Reviewer: ```suggestion
We propose a way to do replay-attack protection without enforcing the order of transactions, without requiring the use of nonces. In this way, we can support un-ordered transaction inclusion.
```

## Context

As of today, the nonce value (account sequence number) prevents replay-attack and ensures the transactions from the same sender are included into blocks and executed in sequential order. However it makes it tricky to send many transactions from the same sender concurrently in a reliable way. IBC relayer and crypto exchanges are typical examples of such use cases.

## Decision

> We propose to add a boolean field `unordered` to transaction body to mark "un-ordered" transactions.

Reviewer: Maybe a silly Q, but if we're repurposing the `timeout_height` field, do we need a new boolean field? Can we determine if a tx wants to be unordered or not by simply evaluating if `timeout_height` is non-zero?

Author: We need to restrict the range of timeout_height of unordered tx, but not for ordered ones.

Reviewer: I see. So you're allowing `timeout_height` to be used for it's original purpose and for unordered txs. Makes sense.

Un-ordered transactions will bypass the nonce rules and follow the rules described below instead, in contrary, the default ordered transactions are not impacted by this proposal, they'll follow the nonce rules the same as before.

> When an un-ordered transaction are included into block, the transaction hash is recorded in a dictionary, new transactions are checked against this dictionary for duplicates, and to prevent the dictionary grow indefinitly, the transaction must specify `timeout_height` for expiration, so it's safe to removed it from the dictionary after it's expired.

Reviewer: ```suggestion
When an un-ordered transaction is included into a block, the transaction hash is recorded in a dictionary. New transactions are checked against this dictionary for duplicates, and to prevent the dictionary grow indefinitely, the transaction must specify `timeout_height` for expiration, so it's safe to removed it from the dictionary after it's expired.
```

> The dictionary can be simply implemented as an in-memory golang map, a preliminary analysis shows that the memory consumption won't be too big, for example `32M = 32 * 1024 * 1024` can support 1024 blocks where each block contains 1024 unordered transactions. For safty, we should limit the range of `timeout_height` to prevent very long expiration, and limit the size of the dictionary.

Reviewer: A short `timeout_height` window also ensures a tighter bound on replay protection.

### Transaction Format

```protobuf
message TxBody {
  ...

  boolean unordered = 4; 
}
```

### `DedupTxHashManager`

```golang
// can reduce frequency we check the expiration.
const ExpireCheckInterval = 1

// DedupTxHashManager contains the tx hash dictionary for duplicates checking,
// and expire them when block number progresses.
type DedupTxHashManager struct {
  // tx hash -> expire block number
  // for duplicates checking and expiration

>   hashes map[TxHash]uint64

Reviewer 2: talking about it below

}

func (dtm *DedupTxHashManager) Contains(hash TxHash) (ok bool) {
  dtm.mutex.RLock()
  defer dtm.mutex.RUnlock()

  _, ok = dtm.hashes[hash]
  return
}

func (dtm *DedupTxHashManager) Size() int {
  dtm.mutex.RLock()
  defer dtm.mutex.RUnlock()

  return len(dtm).hashes
}

func (dtm *DedupTxHashManager) Add(hash TxHash, expire uint64) (ok bool) {
  dtm.mutex.Lock()
  defer dtm.mutex.Unlock()

  dtm.hashes[hash] = expire
  return
}

// EndBlock remove expired tx hashes, need to wire in abci cycles.

> func (dtm *DedupTxHashManager) EndBlock(ctx sdk.Context) {

Reviewer 2: how will this run in endblock? could we not do a quick check against current time and remove everything behind it? could be a `DeleteUntil` then it checks and removes all previous hashes. This could be done async as well if we wanted

Author: > how will this run in endblock?

Wire into baseapp.

Yeah, it can run in background, the end blocker could be a trigger.

It could grab a read lock first and iterate to find out expired hashes, then grab write lock to do the deletion, that should have maximum parallel performance

Author: > that sounds good. Since everything under current height wont be touched by the state machine, do we need locks?

we need locks since check tx runs concurrently, and we also added a background purging loop.

  }
}
```

### Ante Handlers

Bypass the nonce decorator for un-ordered transactions.

```golang
func (isd IncrementSequenceDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
  if tx.UnOrdered() {
    return next(ctx, tx, simulate)
  }
  
  // the previous logic
}
```

A decorator for the new logic.

```golang
type TxHash [32]byte

const (
  // MaxNumberOfTxHash * 32 = 128M max memory usage
  MaxNumberOfTxHash = 1024 * 1024 * 4

  // MaxUnOrderedTTL defines the maximum ttl an un-order tx can set
  MaxUnOrderedTTL = 1024
)

type DedupTxDecorator struct {
  m *DedupTxHashManager
}

func (dtd *DedupTxDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
  // only apply to un-ordered transactions
  if !tx.UnOrdered() {
    return next(ctx, tx, simulate)
  }

  if tx.TimeoutHeight() == 0 {
    return nil, errorsmod.Wrap(sdkerrors.ErrLogic, "unordered tx must set timeout-height")
  }

  if tx.TimeoutHeight() > ctx.BlockHeight() + MaxUnOrderedTTL {
    return nil, errorsmod.Wrapf(sdkerrors.ErrLogic, "unordered tx ttl exceeds %d", MaxUnOrderedTTL)
  }

  if !ctx.IsCheckTx() {

>     // a new tx included in the block, add the hash to the dictionary

Author: For self reminding, check for duplicates in deliverTx as well, checktx is not reliable

    if dtd.m.Size() >= MaxNumberOfTxHash {
      return nil, errorsmod.Wrap(sdkerrors.ErrLogic, "dedup map is full")
    }
    dtd.m.Add(tx.Hash(), tx.TimeoutHeight())
  } else {
    // check for duplicates
    if dtd.m.Contains(tx.Hash()) {
      return nil, errorsmod.Wrap(sdkerrors.ErrLogic, "tx is duplicated")
    }
  }

  return next(ctx, tx, simulate)
}
```

### EndBlocker

> Wire up the `EndBlock` method of `DedupTxHashManager` into the application's abci life cycle.

Reviewer: We could simply purge the map during the `FinalizeBlock` hook that app's can set. There are many places we can purge, it can be `EndBlock` too.

Author: maybe `Commit` event, I changed the method name to `OnNewBlock`.

Reviewer: Yes, overall, my point is that I think this is a minor implementation detail in the grand scheme of things 👍

### Start Up

On start up, the node needs to re-fill the tx hash dictionary of `DedupTxHashManager` by scanning `MaxUnOrderedTTL` number of historical blocks for un-ordered transactions.

An alternative design is to store the tx hash dictionary in kv store, then no need to warm up on start up.

## Consequences

### Positive

* Support un-ordered and concurrent transaction inclusion.

### Negative

> - Start up overhead to scan historical blocks.

Reviewer 2: should we use the filesystem to write the known hashes at shutdown? On start we would populate from the file

Author: As long as we write to disk, we need to handle commit, rollbacks the same as the other state machine state, to keep it consistent. So better to be handled in existing store directly.
but we can cache the whole thing in memory on start up, so duplication checking is fast.

Reviewer 2: we do something like this in store, but not sure its exposed to things out side of store currently. The main worry i have is around upgrades, where everyone stops and starts, then the map would be empty. Looking back at the x blocks would be best but since we dont know if everyone has it, we could end up in a bad situation

Author: yeah, we need some support from storage layer.

## Decision

Add an extra nonce lane to support optional un-ordered transaction inclusion, the default lane provides ordered semantic the same as before, the new one will provide un-ordered semantic. The transaction can choose which lane to use.

One of the design goals is to keep minimal overhead and breakage to the existing users who don't use the new feature.

### Transaction Format

It doesn't change the transaction format itself, but re-use the MSB(most significant bit) of the existing 64bits nonce value to identify the lane, `0` being the default ordered lane, `1` being the new un-ordered lane.

### Account State

The new nonce lane needs to add some optional fields to the account state:

```golang
type UnOrderedNonce struct {
  Sequence uint64
  Timestamp uint64  // the block time when the Sequence is updated
  Gaps IntSet
}

type Account struct {
  // default to `nil`, only initialized when the new feature is first used.

>   unorderedNonce *UnOrderedNonce

Author: changed to `UnorderedNonceManager`.

}
```

> The un-ordered nonce state includes a normal sequence value plus the gap values in recent history, the gap set has a maximum capacity to limit the resource usage, when the capacity is reached, the oldest gap value is simply dropped, which means the pending transaction with that value as nonce will not be accepted anymore.

Author: ```suggestion
The un-ordered nonce state includes a normal sequence value plus the set of unused(gap) values in recent history, these recorded gap values can be reused by future transactions, after used they are removed from the set and can't be used again, the gap set has a maximum capacity to limit the resource usage, when the capacity is reached, the oldest gap value is removed, which also makes the pending transaction using that value as nonce will not be accepted anymore.
```

> The un-ordered nonce state includes a normal sequence value plus the set of unused(gap) values in recent history, these recorded gap values can be reused by future transactions, after used they are removed from the set and can't be used again, the gap set has a maximum capacity to limit the resource usage, when the capacity is reached, the oldest gap value is removed, which also makes the pending transaction using that value as nonce will not be accepted anymore.

Reviewer: We use a bitmap, right? Can you explain rather, what a set bit means and what an unset bit means?

Author: yes, mentioned now that set bits represents the gaps.

Reviewer 3: The gaps were not used before, so "...can be reused..." seems wrong. Maybe "...can still be used..."?

### Expiration

It would be good to expire the gap nonces after certain timeout is reached, to mitigate the risk that a middleman intercept an old transaction and re-execute it in a longer future, which might cause unexpected result.

### Prototype Implementation

The prototype implementation use a roaring bitmap to record these gap values.

```golang
const (
  MSBCheckMask = 1 << 63
  MSBClearMask = ^(1 << 63)
)

// CheckNonce switches to un-ordered lane if the MSB of the nonce is set.
func (acct *Account) CheckNonce(nonce uint64, blockTime uint64) error {
  if nonce & MSBCheckMask != 0 {
    nonce &= MSBClearMask

    if acct.unorderedNonce == nil {
      acct.unorderedNonce = NewUnOrderedNonce()
    }
    return acct.unorderedNonce.CheckNonce(nonce, blockTime)
  }

  // current nonce logic
}
```

```golang
const (
  // GapsCapacity is the capacity of the set of gap values.
  GapsCapacity = 1024
  // MaxGap is the maximum gaps a new nonce value can introduce
  MaxGap = 1024
  // GapsExpirationDuration is the duration in seconds for the gaps to expire
  GapsExpirationDuration = 60 * 60 * 24
)

// getGaps returns the gap set, or create a new one if it's expired

> func(u *UnOrderedNonce) getGaps(blockTime uint64) *IntSet {

Reviewer 3: Nitpicking, `u` is not a nonce, but a nonce manager, or nonce gap manager.

Author: done, changed to `UnorderedNonceManager`

        return errors.New("nonce is occupied or expired")
      }

      gapSet.Remove(nonce)
      u.Gaps = *gapSet
  }
  return nil
}

// IntSet is a set of integers with a capacity, when capacity reached, drop the smallest value
type IntSet struct {
  capacity int
  bitmap roaringbitmap.BitMap
}

func NewIntSet(capacity int) *IntSet {
  return &IntSet{
    capacity: capacity,
    bitmap: *roaringbitmap.New(),
  }
}

func (is *IntSet) Add(n uint64) {
  if is.bitmap.GetCardinality() >= is.capacity {
    // drop the smallest one
    is.bitmap.Remove(is.bitmap.Minimal())
  }

  is.bitmap.Add(n)
}

// AddRange adds the integers in [rangeStart, rangeEnd) to the bitmap.
func (is *IntSet) AddRange(start, end uint64) {
  n := end - start
  if is.bitmap.GetCardinality() + n > is.capacity {
    // drop the smallest ones until the capacity is not exceeded
    toDrop := is.bitmap.GetCardinality() + n - is.capacity
    for i := uint64(0); i < toDrop; i++ {
      is.bitmap.Remove(is.bitmap.Minimal())
    }
  }

  is.bitmap.AddRange(start, end)
}

func (is *IntSet) Remove(n uint64) {
  is.bitmap.Remove(n)
}

func (is *IntSet) Contains(n uint64) bool {
  return is.bitmap.Contains(n)
}
```

> ## Consequences

Reviewer 2: should we break the auth.info struct to add unorderd nonces, then 0 would signify usage of the other one?

> ## Consequences

Author: just added the `unordered` boolean field to enable this feature.

> # ADR 069: Un-Ordered Nonce Lane

Author: thanks, changed.

> # ADR 069: Un-Ordered Nonce Lane

Reviewer: @julienrbrt stole the coveted 69 slot :p

## Decision

Add an extra un-ordered nonce lane to support un-ordered transaction inclusion, it works at the same time as the default ordered lane, the transaction builder can choose either lane to use.

The un-ordered nonce lane accepts nonce values that is greater than the current sequence number plus 1, which effectively creates "gaps" of nonce values, those gap values are tracked in the account state, and can be filled by future transactions, thus allows transactions to be executed in a un-ordered way.

It also tracks the last block time when the latest nonce is updated, and expire the gaps certain timeout reached, to mitigate the risk that a middleman intercept an old transaction and re-execute it in a longer future, which might cause unexpected result.

The design introducs almost zero overhead for users who don't use the new feature.

### Transaction Format

Add a boolean field `unordered` to the `BaseAccount` message, when it's set to `true`, the transaction will use the un-ordered lane, otherwise the default ordered lane.

```protobuf
message BaseAccount {
  ...
  uint64 account_number = 3;
  uint64 sequence       = 4;
  boolean unordered     = 5;
}

> ```

Reviewer 2: @testinginprod thoughts here?

    return NewIntSet(GapsCapacity)
  }

  return &unm.Gaps
}

// CheckNonce checks if the nonce in tx is valid, if yes, also update the internal state.
func(unm *UnorderedNonceManager) CheckNonce(nonce uint64, blockTime uint64) error {
  switch {
    case nonce == unm.Sequence:
      // special case, the current sequence number must have been occupied
      return errors.New("nonce is occupied")

    case nonce > unm.Sequence:
      // the number of gaps introduced by this nonce value, could be zero if it happens to be `unm.Sequence + 1`
      gaps := nonce - unm.Sequence - 1
      if gaps > MaxGap {
        return errors.New("max gap is exceeded")
      }

      gapSet := acct.getGaps(blockTime)
      // record the gaps into the bitmap
      gapSet.AddRange(unm.Sequence + 1, unm.Sequence + gaps + 1)

      // update the latest nonce
      unm.Gaps = *gapSet
      unm.Sequence = nonce
      unm.Timestamp = blockTime

    default:
      // `nonce < unm.Sequence`, the tx try to use a historical nonce
      gapSet := acct.getGaps(blockTime)
      if !gapSet.Contains(nonce) {
        return errors.New("nonce is occupied or expired")
      }

      gapSet.Remove(nonce)

>       unm.Gaps = *gapSet

Author: no, the `Timestamp` only tracks the block time when the latest nonce value updated.
